### PR TITLE
Add EIP-7916 ProgressiveBitlist specs and tests

### PR DESCRIPTION
**Pull request by etan-status**
This follows up on #4445 to extend EIP-7916 support for bitlists, making the `ProgressiveBitlist` type available to implementations in _features.

- https://eips.ethereum.org/EIPS/eip-7916

### CODE REVIEW

**Comment by etan-status** [comment]
- [consensus-spec-tests-4480-1.6.0-a.3.zip](https://drive.proton.me/urls/8VR6ZMSY6W#S7k8xYokdoKY)

File: tests/generators/runners/ssz_generic_cases/ssz_container.py [hunk 1]
```diff
@@ -150,7 +167,7 @@ def invalid_cases():
                 lambda rng=rng, typ=typ: serialize(
                     container_case_fn(rng, RandomizationMode.mode_max_count, typ)
                 )
-                + b"\xff"
+                + b"\x00"
```
Comment (jtraglia) [inline tests/generators/runners/ssz_generic_cases/ssz_container.py:170]: Hmm why did we make this change?

Comment (etan-status) [reply tests/generators/runners/ssz_generic_cases/ssz_container.py:170]: Because appending a non-00 byte does not invalidate a bitlist field in the container.

File: ssz/simple-serialize.md [hunk 2]
```diff
@@ -245,10 +249,10 @@ have to do one of the following depending on what kind of object it is:
 - Containers follow the same principles as vectors, with the difference that
   there may be fixed-size objects in a container as well. This means the
   `fixed_parts` data will contain offsets as well as fixed-size objects.
-- In the case of bitlists, the length in bits cannot be uniquely inferred from
-  the number of bytes in the object. Because of this, they have a bit at the end
-  that is always set. This bit has to be used to infer the size of the bitlist
-  in bits.
+- In the case of bitlists/progressive bitlists, the length in bits cannot be
```
Comment (leolara) [inline ssz/simple-serialize.md:252]: @etan-status  Does this mean that they always end with b10000... until aligned with a byte? Like the last 1 is not part of the bitlist?

Comment (leolara) [reply ssz/simple-serialize.md:252]: If this is the case, we need to think of the corner cases in this case.

For example a bitlist that finishes in the 7th bit with 0 or 1. Also in the 8th bit, 15th and 16th. Like I guess on 7th and 15th bit, there should be a 1 on the last bit and that is and I guess that in 8th and 16th bit it should be follow by 0x80. A wrong implementation of this could fail in this close to the byte boundary case. Even better to extend this to also at 9th and 17th bit.

Comment (etan-status) [reply ssz/simple-serialize.md:252]: Yes, that is the same behaviour as for the Bitlist with bounded capacity: In the serialization, the final `1` bit is not part of the value, and solely indicates the exact bit-count present in the bitlist (to distinguish absent and 0 bits).

Comment (etan-status) [reply ssz/simple-serialize.md:252]: > If this is the case, we need to think of the corner cases in this case.

The existing tests in `ssz_bitlist.py` work by generating a random `BitList` with `test_limit` capacity & content, and then ask the parser to parse it with a `typ_limit` instead.

For example, (1, 2) generates a `BitList[2]` with 2 items, then asks the test runner to parse it with a `BitList[1]` type and assert that it fails to parse.

```python
    for typ_limit, test_limit in [
        (1, 2),
        (1, 8),
        (1, 9),
        (2, 3),
        (3, 4),
        (4, 5),
        (5, 6),
        (8, 9),
        (32, 64),
        (32, 33),
        (512, 513),
    ]:
```

For `ssz_progressive_bitlist.py`, those cases don't apply, because there is no capacity limit. Hence, the only invalid data is the one ending in a zero byte, which is covered by the provided tests.

Comment (leolara) [reply ssz/simple-serialize.md:252]: I am talking about valid case tests.

We should test with endings close to the byte boundary with the last bit being 0 and 1. We are testing that other implementations do it correctly in this cases where there could be a misunderstading about the spec.

for example data: bxxxxxxx1 should correspond with bxxxxxxx1 b10000000 and bxxxxxxx0 to  bxxxxxxx0 b10000000  in the serialization if I understand correctly. These are cases where the implementer could make a mistake more commonly.

Comment (etan-status) [reply ssz/simple-serialize.md:252]: Ah, the tests also contain a metadata yaml file that contains the expected hash_tree_root, and also the expected value in YAML. If the value is not parsed correctly, those values won't match.

Nevertheless, I extended bitlist tests to cover byte boundaries more thoroughly.

Comment (leolara) [reply ssz/simple-serialize.md:252]: I think random values for each bit length is not enough, at byte boundaries we need to include both a test that ends in 0 and a test that ends in 1. Because a wrong implementation could make a bad assumption there.

**Perhaps for each bit length you can include a test that is random with the last bit set to 0 and another test that is random with the last bit set to 1. This is given that the algorithm decides the length based on the last 1.**

I don't think random modes are relevant, but leave them as it is extra, but here the error can be exactly on how the last bit is handled.

Comment (etan-status) [reply ssz/simple-serialize.md:252]: I have updated the tests so that the final bit is forced in some variations.

**Comment by leolara** [comment]
LGTM 🚀 

For me this is a super-set of what we need, so it is ok, but perhaps would be enough instead of having `variation`. To do:

```python
 yield (
                    f"progbitlist_{mode.to_name()}_{length}_{variation}",
                    valid_test_case(
                        lambda rng=rng,
                        mode=mode,
                        length=length: progressive_bitlist_case_fn(
                            rng, mode, length, force_final_bit=True
                        )
                    ),
              )
              
 yield (
                    f"progbitlist_{mode.to_name()}_{length}_{variation}",
                    valid_test_case(
                        lambda rng=rng,
                        mode=mode,
                        length=length: progressive_bitlist_case_fn(
                            rng, mode, length, force_final_bit=False
                        )
                    ),
              )
```

But up to you :-)

**Comment by leolara** [comment]
@jtraglia do you want to merge it?

**Comment by etan-status** [comment]
That would double the test vector count, would you still prefer that?

**Comment by leolara** [comment]
> That would double the test vector count, would you still prefer that?

I think it would be less vectors because there would not be `variation` loop, or am I missing something?

**Comment by jtraglia** [comment]
I think I would like to merge the current version of this PR. Let's do it 😄

**Comment by etan-status** [comment]
> I think it would be less vectors because there would not be `variation` loop, or am I missing something?

Ah, you mean instead of the variations. That's true, but I'm not sure if we should reduce number of tests as a side effect of adding new feature 😅
